native/src/init/rootdir.cpp

```diff
@@ -1,23 +1,86 @@
 #include <sys/mount.h>
 #include <libgen.h>
+#include <list>
 
 #include <magisk.hpp>
 #include <base.hpp>
+#include <pthread.h>
 
 #include "init.hpp"
 #include "magiskrc.inc"
 
 using namespace std;
 
 static vector<string> rc_list;
+using zygote_rc_entry = tuple<string, string, string, sFILE>;
+pthread_cond_t hijack_cond = PTHREAD_COND_INITIALIZER;
+pthread_mutex_t hijack_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static void *hijack_zygote_rc(void *arg) {
+    static auto injected = false;
+    auto &[src, dest, tmp_dir, src_file] = *static_cast<zygote_rc_entry *>(arg);
+    auto dest_file = xopen_file(dest.data(), "w");
+    LOGD("Hijacked [%s]\n", src.data());
+    file_readline(false, src_file.get(), [&dest_file, tmp_dir=tmp_dir.data()] (std::string_view line) -> bool {
+        if (!injected && line.starts_with("service zygote ")) {
+            injected = true;
+            fprintf(dest_file.get(), "%s", line.data());
+            fprintf(dest_file.get(), "    onrestart exec %s/magisk --zygote-restart\n", tmp_dir);
+            return true;
+        }
+        fprintf(dest_file.get(), "%s", line.data());
+        return true;
+    });
+    if (injected) {
+        pthread_mutex_lock(&hijack_mutex);
+        pthread_cond_signal(&hijack_cond);
+        pthread_mutex_unlock(&hijack_mutex);
+        pthread_cond_wait(&hijack_cond, &hijack_mutex);
+    }
+    return nullptr;
+}
 
-static void patch_init_rc(const char *src, const char *dest, const char *tmp_dir) {
-    FILE *rc = xfopen(dest, "we");
-    if (!rc) {
-        PLOGE("%s: open %s failed", __PRETTY_FUNCTION__, src);
-        return;
+static bool hijack_initrc(const char *dir, const char *tmp_dir) {
+    gen_rand_str(nullptr, 0); // init rand
+    xmkdir(INITRCMOCK, 0);
+    char src[PATH_MAX] = {0};
+    char dest[PATH_MAX] = {0};
+    // no slash here to avoid double slash
+    ssprintf(src, sizeof(src), "%sinit.rc", dir);
+    ssprintf(dest, sizeof(dest), "%s/init.rc", INITRCMOCK);
+    LOGD("Hijacking [%s]\n", src);
+    mkfifo(dest, 0644);
```

Comment:
Why do we have to use FIFO here?

```suggestion
    // Mock init.rc with our FIFO so we can change its content
    mkfifo(dest, 0644);
```